libopae++/samples/hello_fpga-4.cpp

```diff
@@ -0,0 +1,143 @@
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#include "opae/cxx/accelerator.h"
+#include "opae/cxx/buffers.h"
+#include <uuid/uuid.h>
+#include <algorithm>
+#include <iostream>
+#include <chrono>
+#include <thread>
+
+using namespace opae::fpga::resources;
+using namespace opae::fpga::application;
+
+static const char* NLB0_AFUID = "D8424DC4-A4A3-C413-F89E-433683F9040B";
+static const uint64_t CL = 64;
+static const uint64_t KB = 1024;
+static const uint64_t MB = KB * 1024;
+static const uint64_t GB = MB * 1024;
+static const uint64_t LOG2_CL = 6;
+static const size_t LPBK1_DSM_SIZE = 4 * KB;
+static const size_t LPBK1_BUFFER_SIZE = 1 * KB;
+static const size_t LPBK1_BUFFER_ALLOCATION_SIZE = 4 * KB;
+static const uint32_t DSM_STATUS_TEST_COMPLETE = 0x40;
+
+static inline uint64_t cacheline_aligned_addr(uint64_t num) {
+  return num >> LOG2_CL;
+}
+
+
+struct nlb0 : public component
+{
+  typedef std::shared_ptr<nlb0> ptr_t;
+
+  nlb0(accelerator::ptr_t accel)
+    : component(accel)
+    , dsm_addr(this)
+    , src_addr(this)
+    , dst_addr(this) 
+    , num_lines(this)
+    , ctl(this)
+    , cfg(this)
+    , dsm(this)
+    , inp(this)
+    , out(this)
+  {
+
+  }
+
+  virtual ~nlb0() {}
+
+  csr64<0x0110> dsm_addr;
+  csr64<0x0120> src_addr;
+  csr64<0x0128> dst_addr;
+  csr32<0x0130> num_lines;
+  csr32<0x0138> ctl;
+  csr64<0x0140> cfg;
+  shared_buffer<LPBK1_BUFFER_SIZE>  dsm;
+  shared_buffer<LPBK1_BUFFER_ALLOCATION_SIZE>  inp;
+  shared_buffer<LPBK1_BUFFER_ALLOCATION_SIZE>  out;
+
+
+private:
+
+};
+
+int main(int argc, char* argv[]) {
+  // look for accelerator with NLB0_AFUID
+  properties filter;
+  filter.guid.parse(NLB0_AFUID);
+  // Call the accelerator::run static function with the following args:
+  // 1. The filter used to locate the accelerator
+  // 2. The flags used when opening the accelerator
+  // 3. The size of the initial buffer - more can be requested from accel
+  // 4. The callback called once an accelerator is found and opened
+  //    This callback takes as arguments the accelerator and the buffer allocated
+  return accelerator::run<nlb0>(filter,
```

Comment:
I'm a bit lost on intuitive semantics here - so `nlb0` is a `component`, which is used as the template class for an `accelerator` class, which has a static method `run()` that creates the accelerator object, enumerates the accelerator based on the `filter`, and calls the provided callback function, right?

That callback then references attributes of the component `nlb0`, which it can because the accelerator was instantiated with `nlb0` as a template. Correct?

Reply:
`accelerator::run` is a static function that:
* uses function arguments to
  * enumerate and open a resource
  * allocate a shared buffer
* uses template argument to
  * create an instance of the class
     * the `component` constructor is called which saves the accelerator and shared buffer
     * the `nlb0` constructor and initialization list is called which
       * initialize CSR members and shared_buffer members to use the underlying accelerator or already allocated buffer


The developer of `nlb0` would only have to specify:
* csr map
* buffers needed
* algorithm to run with the accelerator (either the callback or implementation of `run`)

Reply:
Thanks. How would events work with this pattern? Would we also need to pass events we're interested in through the `run()` function arguments?